### PR TITLE
Consistently default to --lib --bins when --all-targets is not given

### LINKED ISSUE #5134: `--all-targets` doesn't work and contradicts the documentation
~~~~
cargo$ git grep all-targets
src/bin/bench.rs:    --all-targets                Benchmark all targets (default)
src/bin/build.rs:    --all-targets                Build all targets (lib and bin targets by default)
src/bin/check.rs:    --all-targets                Check all targets (lib and bin targets by default)
src/bin/rustc.rs:    --all-targets            Build all targets (lib and bin targets by default)
src/bin/rustdoc.rs:    --all-targets            Build all targets (default)
src/bin/test.rs:    --all-targets                Test all targets (default)
~~~~

These were originally added in #4400, the problem is it currently doesn't work, "all targets" is in effect no matter what and the flag is totally redundant. This is because `CompileFilter` has a `Default` enum which matches all targets (see `CompileFilter::matches`) and this `Default` enum is used in most cases (see `CompileFilter::new`).

I suggest updating `CompileFilter::matches` to match only lib/bin targets for `CompileFilter::Default` and updating the documentation to match.

Note that doing nothing is *not an option*, the documentation is currently incorrect and at the very least needs to be fixed so every instance above says "X all targets (default)" i.e. all the flags are redundant.

### PR DESCRIPTION
**Pull request by infinity0**
Fixes #5134

### CODE REVIEW

**Comment by infinity0** [comment]
r? @alexcrichton based on #4400

cc @behnam

**Comment by alexcrichton** [comment]
Thanks for the PR! I do think though that `cargo check --all-targets`, for example, checks all benchmarks/tests/etc today. It still does with this patch, right?

I also think that we'd probably want to fix the behavior with `cargo {bench,test}` rather than tweak the defaults, as it's intended to "run all tests" and "run all benchmarks" I believe?

**Comment by infinity0** [comment]
Yes, there is a test that `cargo check --all-targets` checks all lib/bin/example/test/bench targets.

After resolving conflicts and fixing a minor bug from #5152, I've additionally added a commit that makes `cargo {bench,test}` applies to all targets (and then the `--all-targets` flag is redundant), as requested.

**Comment by infinity0** [comment]
The tests are failing but I'm pretty sure it's not a problem with my code. If you manually create the example from `build::explicit_examples` you can see that `--all-targets` is already a bit screwed even in older cargo releases:

~~~~
$ cargo --version
cargo 0.24.0
$ cargo clean && cargo test >/dev/null 2>/dev/null; target/debug/examples/hello
Hello, World!
$ cargo clean && cargo test --all-targets >/dev/null 2>/dev/null; target/debug/examples/hello

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

~~~~

I'll dig further, hints welcome.

**Comment by infinity0** [comment]
It looks like `cargo build --examples` vs `cargo test --examples` will each respectively overwrite `target/debug/examples/goodbye` with the example binary vs a test binary that tests the example file.

Currently many of cargo's own tests implicitly depend on the fact that invocations like `cargo test -v --examples` do *not* build the test-for-the-example but instead builds the example binary, for example `build::explicit_examples`. If my latest commit 1d31d9f is reverted/removed from this PR then the tests will start passing again.

Alternatively, it may be possible to switch `cargo {bench,test}` to "all-targets by default" without changing these other tests, by changing the target path of the test-for-the-example to be something other than `target/debug/examples/goodbye`. However this is also fairly invasive, was not the original intention of my PR, and I'd rather not get dragged into a rabbit hole being forced to fix existing deficiencies of cargo that are unrelated to the issue I was trying to fix. I could file an issue for this however.

TL;DR I'd suggest dropping 1d31d9f from this PR and abandoning the current suggestion to have `cargo {bench,test}` applied to all-targets by default, but file an issue about it detailing what the problem is and how to fix it (i.e. either rewrite several-dozen tests, or rewrite the target-paths for test binaries).

**Comment by infinity0** [comment]
There was a simpler way of achieving what I wanted and I misunderstood how the existing code handles `--all-targets`. I'll open another more minimal PR.
